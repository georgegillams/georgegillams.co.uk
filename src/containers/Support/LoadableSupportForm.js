import React from 'react';
import Loadable from 'react-loadable';

const Loading = () => <div />;

let LoadableSupportForm = () => <div>SUPPORT_FORM</div>;
if (process.env.NODE_ENV !== 'test') {
  LoadableSupportForm = Loadable({
    loader: () => import('components/Forms/SupportForm'),
    loading: Loading,
  });
}

export default LoadableSupportForm;
