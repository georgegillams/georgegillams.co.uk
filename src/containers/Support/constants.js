export const KEY = 'support';
export const LOAD_LINKS = `app/${KEY}/LOAD_LINKS`;
export const CREATE_LINK = `app/${KEY}/CREATE_LINK`;
export const DELETE_LINK = `app/${KEY}/DELETE_LINK`;
