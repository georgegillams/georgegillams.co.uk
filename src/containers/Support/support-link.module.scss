@import '~utils/tokens';

.support-link {
  &__link {
    display: inline-block;
    width: 100%;
    margin-bottom: $spacing-base;
  }

  &__delete-button {
    margin-bottom: $spacing-base;
  }
}
