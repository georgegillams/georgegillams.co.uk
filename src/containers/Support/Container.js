import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { cssModules } from '@george-gillams/components/helpers/cssModules';
import DebugObject from 'components/common/DebugObject';
import LoadingCover from '@george-gillams/components/loading-cover';

import SupportSkeleton from './SupportSkeleton';

import STYLES from './support.scss';
import PageTitle from 'components/common/PageTitle';
import Paragraph from '@george-gillams/components/paragraph';
import LoadableSupportForm from './LoadableSupportForm';
import Section from '@george-gillams/components/section';
import SupportLink from './SupportLink';
import Button from 'components/common/Button';
import ErrorDisplay from 'components/common/ErrorDisplay';
import PageContainer from 'components/common/PageContainer';

const getClassName = cssModules(STYLES);

const Support = props => {
  const { loadLinks, createLink, deleteLink, supportState, authenticatorState, className } = props;

  const [newLink, setNewLink] = useState(null);

  useEffect(() => {
    loadLinks();

    const interval = setInterval(() => loadLinks(), 30000);

    const cleanUp = () => {
      clearInterval(interval);
    };
    return cleanUp;
  }, []);

  let supportLinks = null;
  if (supportState.links && supportState.links.map) {
    supportLinks = supportState.links;
  }

  const noSupport = !!supportLinks && supportLinks.length === 0;
  const admin = !!(authenticatorState.user && authenticatorState.user.admin);

  return (
    <PageContainer bottomPadding>
      <DebugObject
        debugTitle="Support"
        debugObject={{
          loadLinks,
          createLink,
          deleteLink,
          supportState,
          authenticatorState,
          className,
        }}
      />
      <PageTitle name="Support">
        <>
          <div className={getClassName('support__controls')}>
            <Button onClick={() => loadLinks()} disabled={supportState.loadingLinks}>
              Reload
            </Button>
          </div>
          <ErrorDisplay message="The support session failed to load" error={supportState.loadLinksError} />
          <LoadingCover loadingSkeleton={SupportSkeleton} loading={!supportLinks}>
            <>
              {noSupport && (
                <Paragraph className={getClassName('support__error')}>
                  No support session is currently active.
                </Paragraph>
              )}
              {supportLinks &&
                supportLinks.map(sL => <SupportLink key={sL.name} admin={admin} link={sL} deleteLink={deleteLink} />)}
            </>
          </LoadingCover>
          {admin && (
            <Section name="ADD LINK">
              <ErrorDisplay message="Failed to add link" error={supportState.createLinkError} />
              <ErrorDisplay message="Failed to delete link" error={supportState.deleteLinkError} />
              <LoadableSupportForm
                link={newLink || {}}
                onDataChanged={setNewLink}
                onSubmit={() => createLink(newLink)}
                disabled={supportState.creatingLink}
              />
            </Section>
          )}
        </>
      </PageTitle>
    </PageContainer>
  );
};

Support.propTypes = {
  supportState: PropTypes.object.isRequired,
  authenticatorState: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
  loadLinks: PropTypes.func.isRequired,
  createLink: PropTypes.func.isRequired,
  deleteLink: PropTypes.func.isRequired,
  className: PropTypes.string,
};

Support.defaultProps = {
  className: null,
};

export default Support;
