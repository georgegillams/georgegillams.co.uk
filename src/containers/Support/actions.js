import { createRoutine } from 'redux-saga-routines';

import { LOAD_LINKS, CREATE_LINK, DELETE_LINK } from './constants';

export const loadLinks = createRoutine(LOAD_LINKS);
export const createLink = createRoutine(CREATE_LINK);
export const deleteLink = createRoutine(DELETE_LINK);
