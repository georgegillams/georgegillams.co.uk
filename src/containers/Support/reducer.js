import produce from 'immer';

import { loadLinks, createLink, deleteLink } from './actions';

export const initialState = {
  links: null,
  loadingLinks: false,
  loadLinksError: null,

  linkToCreate: null,
  creatingLink: false,
  createLinkError: null,

  linkToDelete: null,
  deletingLink: false,
  deleteLinkError: null,
};

const reducer = (state = initialState, { type, payload }) =>
  produce(state, draft => {
    switch (type) {
      case loadLinks.REQUEST:
        draft.loadingLinks = true;
        draft.loadLinksError = null;
        break;

      case loadLinks.SUCCESS:
        draft.loadingLinks = false;
        draft.links = payload.supportMessages;
        break;

      case loadLinks.FAILURE:
        draft.loadingLinks = false;
        draft.loadLinksError = payload;
        break;

      case createLink.TRIGGER:
        draft.linkToCreate = payload;
        break;

      case createLink.REQUEST:
        draft.creatingLink = true;
        draft.createLinkError = null;
        break;

      case createLink.SUCCESS:
        draft.creatingLink = false;
        break;

      case createLink.FAILURE:
        draft.creatingLink = false;
        draft.createLinkError = payload;
        break;

      case deleteLink.TRIGGER:
        draft.linkToDelete = payload;
        break;

      case deleteLink.REQUEST:
        draft.deletingLink = true;
        draft.deleteLinkError = null;
        break;

      case deleteLink.SUCCESS:
        draft.deletingLink = false;
        break;

      case deleteLink.FAILURE:
        draft.deletingLink = false;
        draft.deleteLinkError = payload;
        break;
    }
  });

export default reducer;
