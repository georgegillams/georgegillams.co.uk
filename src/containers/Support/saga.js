import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'utils/common/request';
import apiStructure from 'helpers/apiStructure';

import { selectState } from './selectors';
import { loadLinks, createLink, deleteLink } from './actions';

export function* doLoadLinks() {
  const requestURL = apiStructure.loadSupport.fullPath;

  try {
    yield put(loadLinks.request());

    const result = yield call(request, requestURL, {
      method: 'GET',
    });

    if (result.error) {
      yield put(loadLinks.failure(result));
    } else {
      yield put(loadLinks.success(result));
    }
  } catch (err) {
    yield put(loadLinks.failure(err));
  } finally {
    yield put(loadLinks.fulfill());
  }
}

export function* doCreateLink() {
  const currentState = yield select(selectState());
  const { linkToCreate } = currentState;
  const requestURL = apiStructure.createSupport.fullPath;

  try {
    yield put(createLink.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(linkToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(createLink.failure(result));
    } else {
      yield put(createLink.success(result));
    }
    yield put(loadLinks.trigger());
  } catch (err) {
    yield put(createLink.failure(err));
  } finally {
    yield put(createLink.fulfill());
  }
}

export function* doDeleteLink() {
  const currentState = yield select(selectState());
  const { linkToDelete } = currentState;
  const requestURL = apiStructure.deleteSupport.fullPath;

  try {
    yield put(deleteLink.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(linkToDelete),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(deleteLink.failure(result));
    } else {
      yield put(deleteLink.success(result));
    }
    yield put(loadLinks.trigger());
  } catch (err) {
    yield put(deleteLink.failure(err));
  } finally {
    yield put(deleteLink.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadLinks.TRIGGER, doLoadLinks);
  yield takeLatest(createLink.TRIGGER, doCreateLink);
  yield takeLatest(deleteLink.TRIGGER, doDeleteLink);
}
