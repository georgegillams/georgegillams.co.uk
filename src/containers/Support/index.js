import { memo } from 'react';

import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { loadLinks, createLink, deleteLink } from './actions';
import { selectState as selectAuthenticatorState } from 'containers/common/Authenticator/selectors';
import { selectState } from './selectors';
import Support from './Container';

import injectSaga from 'utils/common/redux/inject-saga';
import injectReducer from 'utils/common/redux/inject-reducer';

import { KEY } from './constants';
import saga from './saga';
import reducer from './reducer';

const mapStateToProps = createStructuredSelector({
  authenticatorState: selectAuthenticatorState(),
  supportState: selectState(),
});

export function mapDispatchToProps(dispatch) {
  return {
    loadLinks: payload => dispatch(loadLinks(payload)),
    createLink: payload => dispatch(createLink(payload)),
    deleteLink: payload => dispatch(deleteLink(payload)),
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);
const withSaga = injectSaga({ key: KEY, saga });
const withReducer = injectReducer({ key: KEY, reducer });

export default compose(withSaga, withReducer, withConnect, memo)(Support);
