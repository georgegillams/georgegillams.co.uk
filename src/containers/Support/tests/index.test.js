import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { initialState as initialAuthenticatorState } from 'containers/common/Authenticator/reducer';
import { initialState } from '../reducer';

import configureStore from 'utils/common/redux/configure-store';

import SupportIndex from '../index';
import Support from '../Container';

const testLinks = [
  { url: 'URL_1', description: 'description 1', name: 'name 1' },
  { url: 'URL_2', description: 'description 2', name: 'name 2' },
];

describe('<Support />', () => {
  let store;
  const spy = jest.fn();

  beforeAll(() => {
    store = configureStore({});
  });

  it('should render correctly - index', () => {
    const { container } = render(
      <Provider store={store}>
        <SupportIndex />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with initial state', () => {
    const { container } = render(
      <Provider store={store}>
        <Support
          loadLinks={spy}
          createLink={spy}
          deleteLink={spy}
          supportState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly when loading', () => {
    const { container } = render(
      <Provider store={store}>
        <Support
          loadLinks={spy}
          createLink={spy}
          deleteLink={spy}
          supportState={{
            ...initialState,
            loadingLinks: true,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with links', () => {
    const { container } = render(
      <Provider store={store}>
        <Support
          loadLinks={spy}
          createLink={spy}
          deleteLink={spy}
          supportState={{
            ...initialState,
            links: testLinks,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with loadError', () => {
    const { container } = render(
      <Provider store={store}>
        <Support
          loadLinks={spy}
          createLink={spy}
          deleteLink={spy}
          supportState={{
            ...initialState,
            loadLinksError: { error: 'some_error', errorMessage: 'some error message' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with admin', () => {
    const { container } = render(
      <Provider store={store}>
        <Support
          loadLinks={spy}
          createLink={spy}
          deleteLink={spy}
          supportState={{
            ...initialState,
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { admin: true },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with admin and createLinkError', () => {
    const { container } = render(
      <Provider store={store}>
        <Support
          loadLinks={spy}
          createLink={spy}
          deleteLink={spy}
          supportState={{
            ...initialState,
            createLinkError: { error: 'some_error', errorMessage: 'Error creating link' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { admin: true },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with admin and deleteLinkError', () => {
    const { container } = render(
      <Provider store={store}>
        <Support
          loadLinks={spy}
          createLink={spy}
          deleteLink={spy}
          supportState={{
            ...initialState,
            deleteLinkError: { error: 'some_error', errorMessage: 'Error deleting link' },
          }}
          authenticatorState={{
            ...initialAuthenticatorState,
            user: { admin: true },
          }}
        />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
