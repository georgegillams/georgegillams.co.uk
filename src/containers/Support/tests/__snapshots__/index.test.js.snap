// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Support /> should render correctly - index 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy button__outer--disabled"
      disabled=""
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <div
    class="skeleton__outer skeleton__card support__body-skeleton"
  >
    <div
      class="skeleton__shimmer"
      style="margin-left: -0px;"
    />
  </div>
</div>
`;

exports[`<Support /> should render correctly when loading 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy button__outer--disabled"
      disabled=""
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <div
    class="skeleton__outer skeleton__card support__body-skeleton"
  >
    <div
      class="skeleton__shimmer"
      style="margin-left: -0px;"
    />
  </div>
</div>
`;

exports[`<Support /> should render correctly with admin 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy"
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <div
    class="skeleton__outer skeleton__card support__body-skeleton"
  >
    <div
      class="skeleton__shimmer"
      style="margin-left: -0px;"
    />
  </div>
  <div
    class="section__outer"
  >
    <h2
      class="section__heading"
      id="add-link"
    >
      ADD LINK
    </h2>
    <div>
      SUPPORT_FORM
    </div>
  </div>
</div>
`;

exports[`<Support /> should render correctly with admin and createLinkError 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy"
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <div
    class="skeleton__outer skeleton__card support__body-skeleton"
  >
    <div
      class="skeleton__shimmer"
      style="margin-left: -0px;"
    />
  </div>
  <div
    class="section__outer"
  >
    <h2
      class="section__heading"
      id="add-link"
    >
      ADD LINK
    </h2>
    <span
      class="paragraph"
    >
      <span
        class="error"
      >
        Failed to add link
        <br />
        Error creating link
      </span>
    </span>
    <div>
      SUPPORT_FORM
    </div>
  </div>
</div>
`;

exports[`<Support /> should render correctly with admin and deleteLinkError 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy"
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <div
    class="skeleton__outer skeleton__card support__body-skeleton"
  >
    <div
      class="skeleton__shimmer"
      style="margin-left: -0px;"
    />
  </div>
  <div
    class="section__outer"
  >
    <h2
      class="section__heading"
      id="add-link"
    >
      ADD LINK
    </h2>
    <span
      class="paragraph"
    >
      <span
        class="error"
      >
        Failed to delete link
        <br />
        Error deleting link
      </span>
    </span>
    <div>
      SUPPORT_FORM
    </div>
  </div>
</div>
`;

exports[`<Support /> should render correctly with initial state 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy"
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <div
    class="skeleton__outer skeleton__card support__body-skeleton"
  >
    <div
      class="skeleton__shimmer"
      style="margin-left: -0px;"
    />
  </div>
</div>
`;

exports[`<Support /> should render correctly with links 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy"
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <div
    class="subsection__outer"
  >
    <h3
      class="subsection__heading"
      id="name-1"
    >
      name 1
    </h3>
    <span
      class="paragraph support-link__link"
    >
      <a
        aria-label="URL_1 (opens in a new tab)"
        class="text-link text-link--text-link text-link--no-padding text-link--inline"
        href="URL_1"
        rel="noopener noreferrer"
        target="_blank"
      >
        URL_1
        <div
          class="text-link__icon-container"
        >
          <svg
            class="icon text-link__icon"
            height="18"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12.5 11.3c.4.4 1 .4 1.4 0L19 6.2V9c0 .6.4 1 1 1s1-.4 1-1V4c0-.6-.4-1-1-1h-5c-.6 0-1 .4-1 1s.4 1 1 1h2.4l-4.9 4.9c-.4.4-.4 1 0 1.4zM19 19c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V7c0-1.1.9-2 2-2h7v2H5v12h12v-7h2"
            />
          </svg>
        </div>
      </a>
      <br />
      description 1
    </span>
  </div>
  <div
    class="subsection__outer"
  >
    <h3
      class="subsection__heading"
      id="name-2"
    >
      name 2
    </h3>
    <span
      class="paragraph support-link__link"
    >
      <a
        aria-label="URL_2 (opens in a new tab)"
        class="text-link text-link--text-link text-link--no-padding text-link--inline"
        href="URL_2"
        rel="noopener noreferrer"
        target="_blank"
      >
        URL_2
        <div
          class="text-link__icon-container"
        >
          <svg
            class="icon text-link__icon"
            height="18"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12.5 11.3c.4.4 1 .4 1.4 0L19 6.2V9c0 .6.4 1 1 1s1-.4 1-1V4c0-.6-.4-1-1-1h-5c-.6 0-1 .4-1 1s.4 1 1 1h2.4l-4.9 4.9c-.4.4-.4 1 0 1.4zM19 19c0 1.1-.9 2-2 2H5c-1.1 0-2-.9-2-2V7c0-1.1.9-2 2-2h7v2H5v12h12v-7h2"
            />
          </svg>
        </div>
      </a>
      <br />
      description 2
    </span>
  </div>
</div>
`;

exports[`<Support /> should render correctly with loadError 1`] = `
<div>
  <h1
    class="page-title__heading"
  >
    Support
  </h1>
  <div
    class="support__controls"
  >
    <button
      class="button__outer button__outer--regular button__outer--no-bouncy"
      type="button"
    >
      <span
        class="button__child"
      >
        Reload
      </span>
    </button>
  </div>
  <span
    class="paragraph"
  >
    <span
      class="error"
    >
      The support session failed to load
      <br />
      some error message
    </span>
  </span>
  <div
    class="skeleton__outer skeleton__card support__body-skeleton"
  >
    <div
      class="skeleton__shimmer"
      style="margin-left: -0px;"
    />
  </div>
</div>
`;
